src/lib.rs

```diff
@@ -365,118 +394,76 @@ where
     ///
-    /// Basic usage:
+    /// let mut args = opts.args();
     ///
+    /// assert_eq!(args.next(), Some("one"));
+    /// assert_eq!(args.next(), Some("two"));
+    /// assert_eq!(args.next(), None);
     /// ```
-    /// use getargs::{Error, Opt, Options, Result};
-    /// let args = ["-a", "1", "3.5", "foo"];
-    /// let opts = Options::new(&args);
-    /// assert_eq!(opts.next(), Ok(Some(Opt::Short('a'))));
-    /// assert_eq!(opts.next(), Ok(None));
-    /// let arg: Result<Option<i32>> = opts.arg();
-    /// assert_eq!(arg, Ok(Some(1)));
-    /// assert_eq!(opts.arg::<f64>(), Ok(Some(3.5)));
-    /// assert_eq!(opts.arg::<i32>(), Err(Error::InvalidArg {
-    ///     desc: "invalid digit found in string".to_string(),
-    ///     value: "foo",
-    /// }));
-    /// ```
-    pub fn arg<T>(&self) -> Result<Option<T>>
-    where
-        T: FromStr,
-        T::Err: Display,
-    {
-        let arg = match self.arg_str() {
-            Some(s) => s.as_ref(),
-            None => return Ok(None),
-        };
-        match T::from_str(arg) {
-            Ok(v) => Ok(Some(v)),
-            Err(e) => Err(Error::InvalidArg {
-                desc: format!("{}", e),
-                value: arg,
-            }),
-        }
+    pub fn args<'opts>(&'opts mut self) -> ArgIterator<'opts, 'str, I> {
+        ArgIterator { inner: self }
     }
 
-    /// Retrieves the rest of the positional arguments, after the
-    /// options have been parsed.
-    ///
-    /// This method returns the list of arguments after the parsed
-    /// options.
-    ///
-    /// # Panics
-    ///
-    /// This method panics if the option parsing is not yet complete;
-    /// that is, it panics if [`next`](#method.next) has not yet
-    /// returned `None` at least once.
-    ///
-    /// # Examples
-    ///
-    /// Basic usage:
-    ///
-    /// ```
-    /// use getargs::{Opt, Options};
-    /// let args = ["-a", "foo", "bar"];
-    /// let opts = Options::new(&args);
-    /// assert_eq!(opts.next(), Ok(Some(Opt::Short('a'))));
-    /// assert_eq!(opts.next(), Ok(None));
-    /// assert_eq!(opts.args(), &["foo", "bar"]);
-    /// ```
-    pub fn args(&self) -> &'a [S] {
-        let inner = self.inner.borrow();
-        match inner.state {
-            State::Start => &self.args[inner.position..],
-            _ => panic!("called args() while option parsing hasn't finished"),
+    /// "Restarts" options parsing if the iterator has been exhausted.
+    /// This only results in any noticeable effect if the iterator is a
+    /// repeating iterator; otherwise, nothing happens.
+    pub fn restart(&'_ mut self) {
+        self.state = match self.state {
+            State::End => State::Start,
+            _ => panic!("called Options::restart() during an iteration")
         }
     }
+
+    /// Consumes `self` and returns the inner iterator. This discards
+    /// some of the argument-parsing state.
+    pub fn into_inner(self) -> I {
+        self.iter
```

Comment:
Maybe add a check here that the state is valid. Probably should panic under the same conditions that `arg()` panics.